### ISSUE TITLE
Need a grid proxy just to start ganga

### ISSUE THREAD

**Issue by rob-c** [ganga-devs/ganga#410]
This keeps coming up time and time again and is causing real problems for me now in debugging is it known what causes the grid proxy to be initialized right before the bootstrap starts to run?

This should __not__ be happening I would like to see this behaviour gone asap.

It's running with the voms tools rather than with the dirac tools which I want it to run with and for some reason it's not finding a valid proxy in my environment and insists on constructing it's own.

From what I can understand this could be understood as a limitation in the Schema as it's not overloading the default proxy handling methods in the first instance that they're called but does so later on giving a false sense that the proxy is being correctly handled imo.

I've a PR incoming (eta 30min or so) which makes the LCG tools stop requesting a proxy on import, it's painful, it's bad and it causes no end of headaches for me on 2 different setups.

I've had to write this in order to submit a job to the correct VO with the GridPP Dirac.

I'd like to get this into 6.1.19 if it isn't too controversial but if there's discussion I'd _really_ like to see this make it into 6.1.20.

Thanks,

Rob

**Comment by milliams** [comment]
I believe that #298 fixes this exact thing by not requesting a proxy if `GLITE_ENABLE` is set. That PR has been open for a month and a half but no-one has reviewed it yet. That PR is also a requirement for the updated credentials system.

**Comment by rob-c** [comment]
No it doesn't. Not for my usecase where this is explicitly still enabled.

I want to have them available and explicitly obey my instructions in order to make use of the .ini file to submit to GridPP Dirac as the correct user. Either we fix (dramatically change) the Schema  or we delay the construction of the proxy until some moment when it is sane to do so.

**Comment by milliams** [comment]
Ah, well this is the problem with the hack of using `GLITE_SETUP` to enable Dirac. It's hardly surprising that Ganga is acting in unwanted ways.

**Comment by rob-c** [comment]
Are we fundamentally saying that the LCG and DIRAC tools are incompatible in this case?

**Comment by milliams** [comment]
Currently, with the existing credentials system, yes. There's no sensible way to support both running in one Ganga session as far as I am aware.

**Comment by rob-c** [comment]
Can we change the default to not construct these in this case. When I'm on lxplus it does every time by default and it's getting to be a pain.

I'll try and do everything without the ```[LCG]``` system in this case but it's a bit annoying to know that the default recipe for DIRAC is currently incompatible with multi-VO users right now.

**Comment by milliams** [comment]
Currently, *Ganga* is incompatible with mutli-VO users. I agree there's things we can do in the short-term to make things a little nicer but it's mostly waiting on my credentials branch.

**Comment by rob-c** [comment]
This is incompatible to the point that ganga does not work for them. This is a very bad situation to be in imo.
Restricting a user to 1 VO at a time is fine, this is a limitation of the tool but to tell them that we offer no support is just bad.

**Comment by rob-c** [comment]
I'm going to change this to stop LCG from starting GLITE by default. It's messy and it drops the user right into a credential construction on startup. This is bad if they want to use something like DIRAC which won't make use of these tools.

I've also got a cvmfs DIRAC env file which will work with the DIRAC commands as well as a modified .ini file which doesn't abuse GLITE. I'm going to upload this somewhere, are there any good suggestions of where might be a good place in git for posterity?

**Comment by drmarkwslater** [comment]
Again, I'm confused as to what the problem is here. If it's just that Ganga requests a credential on startup then OK, but I say in #411, we'd need to tread carefully to make sure LCG isn't broken and I would actually then say that the new Credentials system would be a much better idea. 

As regards using the GLITE system to create DIRAC credentials, I'm not sure where there is an issue if you are a member of multiple VOs. With the .ini files on cvmfs, all it is used for is to make sure the credential is present and correct without setting up the Dirac environment beforehand (which I have seen will cause hard seg faults in readline.so). If you are happy to create the credential yourself then do so outside of ganga and then just set GLITE_ENABLE to False. Though of course, you can't submit/monitor jobs with different credentials in the same session as this would need the Credentials branch.

If this is all really an issue then I would say the best option is to fast track the credentials stuff and get it in to 6.1.20/6.2 - then all of this goes away as Ganga would have proper multi-VO support.

**Comment by rob-c** [comment]
My complaint was that I was under the impression it was somehow needed for the LCG system to be started up to startup DIRAC, since I've been able to do this without GLITE my complaint goes away. I'm still bugged immensely by the fact that the GLITE is turned on by default and that is requires a proxy which can't be generated using the overloaded proxy management methds because the bootstrap hasn't finished.
The fact this interrupts the bootstrap and therefore the startup of Ganga annoys me as it keep happening when I launch it in a vanilla environment to do a simple task. I think it annoys me because I've never used these tools but these tools keep bugging me for a password on startup.

**Comment by drmarkwslater** [comment]
So I guess I wasn't aware that most of this stuff can be handled by the Dirac backend itself (is that true?) in which case we should just have GLITE_ENABLE set to False in the Dirac.ini package file and everything goes away I think. 

I understand that having vanilla Ganga assume you want to run with GLITE is not the best option, certainly with the death of WMS, but that is due to historical reasons (that being the main reason you would run Ganga) it was chosen as the default so I think we're stuck with it until the new Credentials System is in place.

What I think this does all show is that we should prioritise getting the credentials stuff in ASAP as all this becomes a non-issue then :smile: @milliams could you mark the appropriate PRs for 6.1.20 (if you haven't already!) so we know to review them? Thanks!

**Comment by rob-c** [comment]
I'm starting to become strongly against requiring a proxy in order to startup Ganga just by default. This happens when I run a ```git checkout``` and just try and run the code. This is just an annoyance people can probably live without.

**Comment by rob-c** [comment]
I'd much rather ship 6.1.19 with more sane defaults which is what #413 has become. Can we at least agree what the out of the box DIRAC config should be so that someone with cvmfs available can just use it correctly?

**Comment by rob-c** [comment]
As per #379 asking a user to turn an error message off to make some startup warnings go away just seems odd to me.

**Comment by drmarkwslater** [comment]
If we turn off GLITE by default now, there will be users affected (IPPP, possibly SNO+, etc.). If that's deemed acceptable then OK but I would say the better solution is to push the new credentials system.

**Comment by rob-c** [comment]
In this case why not delay getting the proxy for the GLITE system until after bootstrap to get around the bug that it doesn't allow for the ```voms``` tools to be overwritten.

**Comment by milliams** [comment]
@rob-c Do you have a few reproducible steps showing what you want to happen and what happens instead. If you can provide that then I can have a quick look to see if the current credentials system can be tweaked to work as you want?

**Comment by rob-c** [comment]
I want to generate a proxy as per the DIRAC config and submit a job. The aptly named ```check_proxy``` command in LCG calls for the proxy to be renewed regardless of how valid it is.

Steps to reproduce:
Change the gridproxy commands in the config to be anything and the GLITE system will not be able to run with this as the GLITE system requests for a proxy to be generated on the import of the class.


If GLITE was being run after the schema had finalized then it would work or if the Schema was able to change the default value of the Credentials but it can't before the LCG is imported.
A quick-fix may be to guarantee the import of Credentials is done early enough that it's defaults in the Schema are changed before the LCG system is imported but this requires understanding the order of all of the imports and cross-dependencies and that the Schema doing what we want.

**Comment by drmarkwslater** [comment]
@rob-c I may be doing something wrong but I don't see this, at least with 6.1.18 that's on CVMFS. Is that the version you're trying or is it `develop`? Maybe there's been a bug introduced because I can happily get Ganga to use the `dirac-proxy-*` commands instead of `voms-*` and it definitely generates a proper Dirac proxy...

**Comment by drmarkwslater** [comment]
I've found the problem. @rob-c I'm guessing you are setting the `[defaults_GridProxy]voms` setting in your `.gangarc`? If so, that's what's triggering the use of the `VomsCommand` stuff. If that is left blank then a Voms Proxy isn't created and a basic Grid Proxy is used which uses the `GridCommands` stuff. With those overridden to the dirac commands, everything works.

**Comment by rob-c** [comment]
I've bumped this #411 and #413 to 6.1.20 it can be dealt with in that release cycle.

**Comment by drmarkwslater** [comment]
It seems that you also need to blank `[LCG]VirtualOrganisation` as well. This is trickier because the default is to set this to dteam. I'm not sure why I've not seen this problem before, but there we go. I'll put in a PR to change the default to '' as that should help things.

**Comment by rob-c** [comment]
Does it look reasonable to add some basic proxy caching into the LCG layer to prevent the default GLITE requiring a proxy right upon startup?
This can be achieved with a light modification to #411. On the other hand if we suspect the LCG tools to probably have some problems with getting the proxy during runtime. i.e. not all of the code checks for a proxy before executing commands which require one I will close the PR.
We should probably be helping people who don't want this to turn it off in the wiki somewhere aswell imo. I can imagine people wanting to submit jobs to a local batch for instance getting irritated that they need a grid proxy to do so.

**Comment by milliams** [comment]
As far as I am aware, the LCG tools don't check for the existence of a grid proxy in a reliable way. This is one of the things that precipitated the creation of the new credentials system so that these things can be deferred to the right time.

**Comment by rob-c** [comment]
OK, so once we have a wiki entry helping people turn this off if they happen to have voms tools or others installed but aren't using them I'll close the PR and look forward to 6.3

**Comment by twhyntie** [comment]
Aha! I was going to create a new issue for this (and I've commented on #379) but I _think_ it's covered by this one. If we want to promote Ganga as a gateway tool to the grid for new users (which I personally do, as things stand), it would be great if they weren't presented with a lot of warnings about not having a grid certificate/proxy/etc when they first run Ganga...

**Comment by drmarkwslater** [comment]
Due to the messy state of the code and the new Credentials stuff that's incoming, I think this is best sorted out with @rob-c suggestion in #379 of basically just having a one line warning saying the GLITE system wasn't set up correctly so this functionality is being turned off. I've created an issue to deal with this separately #486

**Comment by rob-c** [comment]
I think this has been resolved by now. We can re-open this if needed
